qcodes/instrument_drivers/yokogawa/GS200.py

```diff
@@ -191,38 +191,43 @@ def __init__(self, name: str, address: str, terminator: str = "\n",
                            set_cmd=self._set_source_mode,
                            vals=Enum('VOLT', 'CURR'))
 
-        # When getting the mode internally in the driver, look up the mode as
-        # recorded by the _cached_mode property, instead of calling source
-        # _mode(). This will prevent frequent VISA calls to the instrument.
-        # Calling _set_source_mode will change the chased value.
-        self._cached_mode = "VOLT"
+        # We need to get the source_mode value here as we cannot rely on the
+        # default value that may have been changed before we connect to the
+        # instrument (in a previous session or via the frontpanel).
+        self.source_mode()
+
+        self.add_parameter('range',
+                           label='Source Range',
+                           get_cmd=':SOUR:RANG?',
+                           set_cmd=partial(self._set_range, self.source_mode()),
+                           vals=Enum(1e-3, 10e-3, 100e-3, 200e-3, 1e0,
+                           10e0, 30e0),
+                           )
 
-        # We want to cache the range value so communication with the instrument
-        # only happens when the set the range. Getting the range always returns
-        # the cached value. This value is adjusted when calling self._set_range.
-        self._cached_range_value: Optional[float] = None
+        # We need to get the range value here as we cannot rely on the
+        # default value that may have been changed before we connect to the
+        # instrument (in a previous session or via the frontpanel).
+        self.range()
 
         self.add_parameter('voltage_range',
                            label='Voltage Source Range',
                            unit='V',
                            get_cmd=partial(self._get_range, "VOLT"),
                            set_cmd=partial(self._set_range, "VOLT"),
-                           vals=Enum(10e-3, 100e-3, 1e0, 10e0, 30e0))
+                           vals=Enum(10e-3, 100e-3, 1e0, 10e0, 30e0),
+                           snapshot_exclude=(True if self.source_mode() ==
+                           'CURR' else False))
 
-        # The driver is initialized in the volt mode. In this mode we cannot
-        # get 'current_range'. Hence the snapshot is excluded.
         self.add_parameter('current_range',
                            label='Current Source Range',
                            unit='I',
                            get_cmd=partial(self._get_range, "CURR"),
                            set_cmd=partial(self._set_range, "CURR"),
                            vals=Enum(1e-3, 10e-3, 100e-3, 200e-3),
-                           snapshot_exclude=True
+                           snapshot_exclude=(True if self.source_mode() ==
+                           'VOLT' else False)
                            )
 
-        # This is changed through the source_mode interface
-        self.range = self.voltage_range
```

Comment:
Rather than replacing this with a new parameter range should be an alias for voltage range if we are in voltage mode and an alias for current range if we are in current mode